axiom/nr_log_events.c

```diff
@@ -8,32 +8,214 @@
+  }
+}
+
+int nr_log_event_wrapped_priority_comparator(const void* a,
+                                             const void* b,
+                                             void* userdata NRUNUSED) {
+  COMPARATOR_NULL_CHECK(a, b);
+
+  return nr_log_event_priority_comparator((const nr_log_event_t*)a,
+                                          (const nr_log_event_t*)b);
+}
+
+/* All log events popping out of the segment heap go through this
+ * function.
+ *
+ * The log event is discarded (thus removed from the log event heap).
+ */
+static void nr_log_event_discard_wrapper(nr_log_event_t* event,
+                                         void* userdata NRUNUSED) {
+  if (NULL == event) {
+    return;
+  }
+
+  nr_log_event_destroy(&event);
+}
+
+nr_log_events_t* nr_log_events_create(size_t max_events) {
+  nr_log_events_t* events;
+
+  if (NR_MAX_LOG_EVENTS_MAX_SAMPLES_STORED < max_events) {
+    return NULL;
+  }
+
+  events = (nr_log_events_t*)nr_zalloc(sizeof(nr_log_events_t));
+  events->events_allocated = max_events;
+  events->events_used = 0;
+  events->events_seen = 0;
+
+  if (0 != max_events) {
+    events->events = nr_minmax_heap_create(
+        max_events, nr_log_event_wrapped_priority_comparator, NULL,
+        (nr_minmax_heap_dtor_t)nr_log_event_discard_wrapper, NULL);
+  } else {
+    events->events = NULL;
+  }
+
+  return events;
+}
 
+void nr_log_events_destroy(nr_log_events_t** events_ptr) {
+  nr_log_events_t* events;
+
+  if (NULL == events_ptr) {
+    return;
+  }
+  events = *events_ptr;
   if (NULL == events) {
-    return false;
+    return;
+  }
+
+  if (NULL != events->events) {
+    nr_minmax_heap_destroy(&events->events);
   }
+
+  nr_realfree((void**)events_ptr);
+}
+
+size_t nr_log_events_max_events(const nr_log_events_t* events) {
+  if (NULL == events)
+    return 0;
+
+  return events->events_allocated;
+}
+
+size_t nr_log_events_number_seen(const nr_log_events_t* events) {
+  if (NULL == events)
+    return 0;
+
+  return events->events_seen;
+}
+
+size_t nr_log_events_number_saved(const nr_log_events_t* events) {
+  if (NULL == events)
+    return 0;
+
+  return events->events_used;
+}
+
+bool nr_log_events_add_event(nr_log_events_t* events,
+                             const nr_log_event_t* event) {
+  bool events_sampled = false;
+  nr_log_event_t* dup_event;
+
   if (NULL == event) {
     return false;
   }
 
-  events_sampled = nr_analytics_events_is_sampling(events);
+  // need to increment this if a valid event was sent
+  // so take care if it here then sanity check other args
+  if (NULL != events)
+    events->events_seen++;
 
-  log_event_json = nr_log_event_to_json(event);
-  log_event = nr_analytics_event_create_from_string(log_event_json);
-  nr_free(log_event_json);
+  // if no event queue exists or size is 0 then event will be dropped
+  if (NULL == events || NULL == events->events
+      || 0 == events->events_allocated) {
+    return true;
+  }
```

Comment:
Style nit global comment.
Inconsistent use of `{}` after one line if statements.
I'd vote for using the `{}` with all the if statements because it makes it clear and makes it easier to add additional lines to the if statement later.

Reply:
Thanks for catching this - I noticed some inconsistencies but didn't catch them all.

Should be fixed in commit [a4b26a8](https://github.com/newrelic/newrelic-php-agent/pull/515/commits/a4b26a8d2106e7d06b164a34f5e7c3081779bf46).

Reply:
Thanks!  One more here, I think: https://github.com/newrelic/newrelic-php-agent/pull/515/files#diff-80d8ae697d75559f2c45cc1e90dd0fa8b210833ed35750748f156f8c651f1cb4R143

Reply:
Doh - fixed in commit [b24a5cf](https://github.com/newrelic/newrelic-php-agent/pull/515/commits/b24a5cfb856bdba70b50583a3c9e47b5a7d89e76).